### ISSUE TITLE
Usability with screen readers

### ISSUE THREAD

**Issue by saulshanabrook** [jupyterlab/jupyterlab#4878]
I am opening this issue to track being able to use JupyterLab easily with a screen reader.  [@SamKacer on gitter](https://gitter.im/jupyterlab/jupyterlab?at=5b4645f31c0f906b1448aa7b) raised the issue that they are not able to edit files in JupyterLab with a screen reader.

We use the CodeMirror editor for the built in file editor and the cell editors, which is not usable with screen readers (https://github.com/jupyter/notebook/issues/1801#issuecomment-315579069 https://github.com/codemirror/CodeMirror/issues/4604).

There is an extension that adds a Monaco based text editor (https://github.com/jupyterlab/jupyterlab-monaco) and [Monaco looks like it has support for screen readers](https://github.com/Microsoft/monaco-editor/wiki/Monaco-Editor-Accessibility-Guide#ui-optimized-for-screen-readers), so using that extension might provide users so better support. 

It might also make sense to add tests for screen reader based navigation. I am not sure if this is doable with selenium or something.

**Comment by SamKacer** [comment]
@saulshanabrook Thanks for informing me about the other editor. I will test it out first and report on its accessibility. If it is good enough I think that is sufficient for screen reader users. If not, I will expand on SR accessibility.

**Comment by SamKacer** [comment]
going back and playing around with google docs, it looks like it handles accessibility with the lattr approach with ARIA live regions. 

here is a document which details using ARIA as well as web accessibility in general:
https://www.w3.org/TR/aria-in-html/

Also, didnt mention, but the menu bar is usable, but only using a SR workaround. The best accessibility of elements like menu items is for them to be focusable using keyboard actions (like tabbing) and activateble by either inheriting from already activatable elements like html native buttons or links, or by scripting the activatability. This allows the SR user to activate the menu item or whatever interactive element by pressing space when the system focus is on the item.

It is also useful to telegraph the role of the elements if they are not native html widgets such as in case  of the menu bar items in jupyter lab, by applying the role attribute with the right value to the div containing the element.

**Comment by ian-r-rose** [comment]
Thanks a ton for your input @SamKacer. Improving the accessibility of JupyterLab (especially with screen readers) is important to us, though our progress on that front has been slow, largely due to lack of expertise on our end.

A couple of months ago we had an accessibility audit with a SR with the UC Berkeley DSP office. JupyterLab is being deployed to students at a large scale at Berkeley, and accessibility is a priority. The conclusions of that meeting were consistent with what you are saying here. In particular, the CodeMirror editor is largely unusable with a SR.

You are right that Monaco is the same underlying editor used in VS Code, and so switching to that would be a great improvement. In fact, that is one of the primary reasons that we want to move to Monaco.

Our intention is for the menus to be more accessible than they apparently are right now. Once they are active I can navigate around them with the arrow keys, but I see now that they are not tab-activatable as it is (ping @sccolbert). Would it be a good place to start to (1) give the top-level menus a tab-index, and (2) add an aria role attribute to them?

Again, thank you very much for your feedback. As we move forwards with this process, we can add more information to this issue (and would welcome further thoughts!).

**Comment by SamKacer** [comment]
@ian-r-rose
No problem. I am pretty excited about the jupyter project from what I hear from my sighted counterparts, but right now as a blind user I cant fully enjoy it. I am interested in exploring data science and jupyter seems to be an excellent platform for such things. Increasingly more visually impaired people are getting into the IT sector and having jupyter lab be accessible would be a massive boon for the VI community.

If the VS code editor could be used in jupyter lab that would be great if it retained all its features and accessibility it has in the desktop version. Looking at the front end for VS code, it looks like the GUI is basically made of html, so hopefully if it was incorporated into jupyter lab, its functionality would remain the same.

Yes, the changes you propose are steps in the right direction, but more would need to be done for the menubar to be considered fully accessible. I need to stress that the ARIA role attr is merely semantic. It allows the SR to inform the user about its semantic usage and so they expect  certain keyboard control featuresto be implemented (i.e.  pressing left or right arrow will move between the items on the menu bar, pressing down, space, or enter while focused on  an item will open the submenu and move the focus to its first item, etc.).

Right now while researching ARIA, I found these two great documents which not only describe using ARIA, but designing accessible webapps in general.

This one I already linked to, but is a great starting point:
https://www.w3.org/TR/aria-in-html/

and here is a doc where among other things it describes the keyboard control conventions for the ARIA roles and example implementations including specifically the menu bar:
https://www.w3.org/TR/wai-aria-practices/#menu

the above link directly  goes to the section about menu bars.

For webapps to be SR accessible, having comprehensive keyboard navigation in place is key. ARIA role attrs help describe to  the SR user  what sort of keyboard interaction convention to expect, but don’t actually implement it. Fortunately, applying ARIA roles and states is very simple, but thankfully keyboard nav which is harder to implement is useful for all users, not just for SR users.

So, as far as the menu bar the latter link comprehensively explains and demonstrates all the changes that would make it 100
% accessible.

Also I will mention an important point, when referring to focus, for focus switching to be useful to a SR it needs to be the system focus and not a virtual focus. Not sure how keyboard nav was planned on being implemented, but I am mentioning this just in case.

anyway, I will be happy to test the changes as they are implemented, but the doc is pretty straightforward as to what needs to be implemented and the keyboard nav can be tested by any user, while ARIA  roles, states, etc. are very simple to apply.

**Comment by SamKacer** [comment]
Hello all,

I have thought of a quick fix, which might be pretty simple to implement and would allow for accessibility of the editors, which is the primary screen reader accessibility problem.

What if the user could toggle a “simple editor mode”, which would replace the code mirror editors with plain html multi line edit fields? I know the advanced editor features such as  autocompletion  wouldn’t be available, but at the current stage, the only accessible feature in CM is copying the contents over to an external accessible editor anyway.

Would this be possible?

**Comment by jasongrout** [comment]
Or it may be simpler to have two notebook renderers. You right-click and "open with" the simplified notebook editor (or you can set it to be the default).

**Comment by vidartf** [comment]
@SamKacer The editor is already an extensible point, so making a third-party extension that implements a minimal editor would be an interesting exercise (not one that I can volunteer for though). See the `IEditorServices` interface in packages/codeeditor for the functionality that would need to be implemented.

**Comment by jmh-aira** [comment]
Hi, I wanted to follow up with th is? has any headway been made in making Jupyter accessible with an SR?  
Thanks.

**Comment by tonyfast** [comment]
crossref #9399

**Comment by isabela-pf** [comment]
Since we're returning to this issue with some further discussion at #9399, I'm taking @SamKacer's advice to follow up on this issue so anyone following this but not newer issues gets updated too. 

Three potential solutions proposed by @jasongrout 
> * CodeMirror 6 (at that time just starting) looked like it would be significantly better for accessibility - I still would love if someone could try out the CodeMirror 6 demo at https://codemirror.net/6/ to test this assumption
> * Monaco - we have an experimental plugin changing the text editor for monaco, but it suffered from other issues, including for a long time monaco wasn't really packaged well for consumption outside of VS Code, and that it was pretty RAM intensive compared to codemirror, which would possibly make long notebooks suffer a lot in our current architecture.
> * We could fall back to just a straight HTML text area, which certainly is accessible, even if it does not offer any coding help.

This issue (and #9399's discussion) were brought up at this week's JupyterLab general team meeting. This isn't a meeting where we came up with solutions, but gave a heads up to more people that there's interest in proposing changes to the code editor, a major part of JupyterLab. Quoting my own summary of the meeting:
> * There is not an agreement on which solution to pursue, but I expected that at this point. I personally think exploring the options is one of the next steps.
> * Out of who attended this week, there wasn’t any outright opposition to proposing changes to the code editor for this reason (yay!).
> * The big takeaway is that this is not a 3.0 change but a  3.x change at the earliest, possibly 4.0. Some people in this meeting did agree this should be one of the first priorities for the next release and that it should get triage tags and appropriate attention as soon as 3.0 is out.

I also want to note that I have this on the agenda to discuss at the JupyterLab accessibility meeting next Wednesday (Dec 16 at 10:15 AM Pacific; on the [Jupyter Zoom channel](https://zoom.us/my/jovyan?pwd=c0JZTHlNdS9Sek9vdzR3aTJ4SzFTQT09)), where I hope to talk more about where this fits on our priority list and the proposed solutions so we are prepared to bring work back to the rest of the team sooner than later. Anyone is welcome to join this meeting and/or other the other accessibility meetings that are happening (every other Wednesday at 10:15 AM Pacific on the same channel). We are working to address the backlog of accessibility issues that we've left untouched for (in my opinion) too long, and these meetings are meant to coordinate those efforts.

**Comment by fcollonval** [comment]
I follow from afar the discussion. And I think there is a important point that has not been mentioned about the editor: the compatibility with non-desktop browsers. This is not per se related to accessibility. But monaco editor is not for mobile browser (as clearly stated in the [project homepage](https://microsoft.github.io/monaco-editor/)) when CodeMirror is. It may be a blocking point when choosing a more accessible editor.
